[Messenger] Fix merging PrototypedArrayNode associative values

svitiashchuk opened the pull request:
Now keys will not be considered as irrelevant.

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #39846, fix #36874
| License       | MIT
| Doc PR        | 

The problem was because in ConfigTree `options` are defined as `arrayNode` and has a `prototype('variable')`. So for combining config together `PrototypedArrayNode::mergeValues()` was called. And there was a trouble - for that `PrototypedArrayNode` `$rightSide` values were added without keys:

`\Symfony\Component\Config\Definition\PrototypedArrayNode::mergeValues`
```php
foreach ($rightSide as $k => $v) {
    // prototype, and key is irrelevant, append the element
    if (null === $this->keyAttribute) {
        $leftSide[] = $v;
        continue;
    }
```

svitiashchuk committed:
[Messenger] Fix merging PrototypedArrayNode associative values

jderusse reviewed (commented):
could you please add a test?

see https://github.com/symfony/symfony/pull/39686/files for similar PR

svitiashchuk commented:
Test added

jderusse commented:
looks like the failed tests is related to your change

svitiashchuk commented:
Yes, it was. Now, I've changed the argument `$name` value of method `ArrayNodeDefinition::useAttributeAsKey` to `'option'`. Because in fixture config `messenger_transports.php` there was a piece:
```
'options' => ['queue' => ['name' => 'Queue']],
```
for which `'name' => 'Queue'` was considered as key for option, and therefore, rewritten the original key `'queue'` with `'Queue'`. Now that problem solved.



P.S. I'm not sure about that, but is that a normal behaviour that original key replaced when `attributeAsKey` is set and original key is present in config simultaneously. Should I create another one issue?

jderusse commented:
Indeed, the attributeAsKey should not be an expected subkey of the array.

OskarStark commented on src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/ConfigurationTest.php:
```diff
@@ -278,6 +278,45 @@ public function testLockMergeConfigs()
+            ],
+        ]);
+
+        $this->assertEquals(
```
```suggestion
        $this->assertSame(
```
?

OskarStark commented:
![CleanShot 2021-01-15 at 15 49 24](https://user-images.githubusercontent.com/995707/104741265-43673b00-5749-11eb-8e5a-0f8a19293f7b.png)

It looks like your committer email is not associated with your Github account 🤔

svitiashchuk commented:
@OskarStark yes, thanks for pointing. Did it in different environment, but now fixed it.

xabbuh reviewed (changes requested):
I am not convinced that this will always work. What if there is a queue where one option is named `option`?

svitiashchuk reviewed (commented):
@xabbuh I've added an `option` key inside test config options array to prove that this will work.

xabbuh commented:
@svityashchuk If you change your example to something like this, you will experience the same behavior:

```php
'options' => [
    'auto_setup' => true,
    'queue' => [
        'name' => 'Queue',
        'option' => 'foo',
    ],
],
```

svitiashchuk commented:
@xabbuh thank you for providing an example. I've got that behaviour. So, in the circumstances I see the only possibility to use another keyword as keyAttribute. Otherwise, I won't be able to pass this check at `\Symfony\Component\Config\Definition\PrototypedArrayNode::mergeValues`:

```php
foreach ($rightSide as $k => $v) {
    // prototype, and key is irrelevant, append the element
    if (null === $this->keyAttribute) {
        $leftSide[] = $v;
        continue;
    }
```

Moreover, now it seems for me like this condition `null === $this->keyAttribute` is not enough to decide to append value as plain array item (I mean, not an associative array element). Therefore, I think we need to add an additional condition, like it is done in `\Symfony\Component\Config\Definition\PrototypedArrayNode::normalizeValue`. There we can see the next lines:
```php
protected function normalizeValue($value)
{
//...
$isAssoc = array_keys($value) !== range(0, \count($value) - 1);
//...
foreach ($value as $k => $v) {
    //...
    if (null !== $this->keyAttribute || $isAssoc) {
        $normalized[$k] = $prototype->normalize($v);
    } else {
        $normalized[] = $prototype->normalize($v);
    }
```

Pay attention that `$isAssoc` variable affects the way normalized value is added to resulting `$normalized` array.
If you wouldn't mind, I will provide corresponding changes.

xabbuh commented:
This sounds like a solution that could work. Let's give that a try. For `5.x` I have pending changes locally that allow to configure list and map nodes explicitly which would allow an easier solution. But that's of course not something we can use here for the bugfix.

svitiashchuk commented:
Now it seems like everything is ok :)
Thanks to my colleague I realized that I had written test in different bundle. Therefore, on Travis CI check it was impossible to test changes that I had provided. So I've removed that test and wrote another one in relevant directory.

I would be very happy if this is merged.
@xabbuh could you please take a look at it?

jderusse commented on src/Symfony/Component/Config/Definition/PrototypedArrayNode.php:
```diff
@@ -304,9 +304,10 @@ protected function mergeValues($leftSide, $rightSide)
             return $rightSide;
         }
 
+        $isAssoc = array_keys($rightSide) !== range(0, \count($rightSide) - 1);
```
What's about using `array_is_list` and requiring the `php81` polyfill?

svitiashchuk replied:
Updated PR with this polyfill.

nicolas-grekas commented on src/Symfony/Component/Config/Definition/PrototypedArrayNode.php:
```diff
@@ -225,7 +225,7 @@ protected function normalizeValue($value)
         $value = $this->remapXml($value);
 
-        $isAssoc = array_keys($value) !== range(0, \count($value) - 1);
+        $isAssoc = !array_is_list($value);
```
maybe `$isList = array_is_list($value);` to prevent the double negation? (same in the other method)

nicolas-grekas reviewed (commented):
Please update the PR title - it's not about Messenger :)

Sscorpion9991 commented:
Hi! Is this PR going to be merged? I have the same issue.

derrabus commented:
Thank you @svityashchuk.
